#!/usr/bin/env Rscript

library(readr)
library(dplyr)
library(cowplot)
library(purrr)
library(stringr)
library(tidyr)


datapath <- '/stor/work/Lambowitz/cdw2854/ecoli_genome/base_indel_table'
indel_tables <- list.files(path = datapath, 
                           pattern = '.tsv',
                           full.names = T)
indel_tables <- indel_tables[grepl('UMI',indel_tables)]

complementary_base <- function(b){
    return (str_replace(b, 'ACTG','TGAC'))
}


read_file <- function(indel_table){
    table_name <- basename(indel_table)
    samplename <- str_replace(table_name,'.tsv','')
    df <- read_tsv(indel_table) %>%
        gather(indel, counts, -mononucleotide, -run_length) %>%
        mutate(strand = ifelse(grepl('rev|neg',indel),'Reverse strand', 'Foward strand')) %>%
        mutate(indel = str_replace_all(indel,'negatve_|positive_|fwd_|rev_','')) %>%
        mutate(mononucleotide = ifelse(strand == 'Reverse strand',chartr('ACTG','TGAC',mononucleotide),mononucleotide)) %>%
        group_by(mononucleotide, run_length, indel) %>%
        summarize(counts = sum(counts)) %>%
        spread(indel, counts) %>%
        gather(indel, counts, -mononucleotide:-aln_count) %>%
        mutate(indel_rate = counts/aln_count) %>%
        mutate(method = ifelse(grepl('nextera',samplename),'Nextera-XT','TGIRT-seq')) %>%
        mutate(samplename = samplename) %>%
        tbl_df 
    return(df)
}

df <- indel_tables %>%
    map_df(read_file) %>%
    filter(method != 'Nextera-XT') %>%
    mutate(indel = str_to_title(indel)) %>%
    tbl_df

p<-ggplot(data = df, 
       aes(color = mononucleotide, 
           x = run_length, 
           #shape = method,
           y = indel_rate)) + 
    geom_jitter(alpha=0.3, size=2) + 
    facet_grid(indel~., scale='free_y')+
    #panel_border()+
    #viridis::scale_color_viridis() +
    labs(color = ' ', x = 'Homopolymer length (nt)',
         y = 'Indel per read per homopolymer',
         shape = '') +
    theme(text = element_text(size=25, face='bold')) +
    theme(legend.key.size = unit(2,'line')) 
    #scale_color_manual(values=c('red','gold','green','blue'))
    
figurename <- str_c(datapath, '/base_mononucleotide.pdf')
ggsave(p, file =  figurename, height = 7, width = 7)
message('plotted: ', figurename)
